doc/sg_seek.8
.TH SG_SEEK "8" "September 2018" "sg3_utils\-1.43" SG3_UTILS
.SH NAME
sg_seek \- send SCSI SEEK, PRE-FETCH(10) or PRE-FETCH(16) command
.SH SYNOPSIS
.B sg_seek
[\fI\-\-10\fR] [\fI\-\-count=NC\fR] [\fI\-\-grpnum=GN\fR] [\fI\-\-help\fR]
[\fI\-\-immed\fR] [\fI\-\-lba=LBA\fR] [\fI\-\-num\-blocks=NUM\fR]
[\fI\-\-pre\-fetch\fR] [\fI\-\-readonly\fR] [\fI\-\-skip=SB\fR]
[\fI\-\-time\fR] [\fI\-\-verbose\fR] [\fI\-\-version\fR]
[\fI\-\-wrap\-offset=WO\fR] \fIDEVICE\fR
.SH DESCRIPTION
.\" Add any additional description here
Sends a SCSI SEEK(10), PRE\-FETCH(10) or PRE\-FETCH(16) command to the
\fIDEVICE\fR. The SEEK command has been obsolete since SBC\-2 (2005) but
still is supported on some hard disks and even some SSDs (solid state
disks). The PRE\-FETCH command can be viewed as SEEK's modern replacement.
Instead of talking about moving the disk heads to the track containing
the sort after LBA, it talks about bringing the sort after LBA (and a
given number of blocks) into the disk's cache. Also the PRE\-FETCH commands
have an IMMED field.
.PP
The PRE\-FETCH commands can report "real" errors but usually they will report
one of two "good" statuses. To do this they return the rarely used CONDITION
MET status. If the number of blocks does actually fit in the cache (when
IMMED=0) or there is enough room in the cache when the command arrives (when
IMMED=1) then a CONDITION MET status is returned. If the requested number of
blocks did not fit (IMMED=0) or would not fit (IMMED=1) then status GOOD
is returned. So if a disk has a large cache and PRE\-FETCH is used sparingly
then the command is more likely to return CONDITION MET than GOOD. This
presents some SCSI sub\-systems with problems as due to its rareness they
mishandle CONDITION MET and treat it as an error (see NOTES section below).
.SH OPTIONS
Arguments to long options are mandatory for short options as well.
.TP
\fB\-T\fR, \fB\-\-10\fR
use a 10 byte cdb command, either SEEK(10) or PRE\-FETCH(10) command. In
the absence of the \fI\-\-pre\-fetch\fR option, the SEEK(10) command is
used. If the \fI\-\-pre\-fetch\fR option is given without this option
then a PRE\-FETCH(16) command is used.
.TP
\fB\-c\fR, \fB\-\-count\fR=\fINC\fR
\fINC\fR is the number of commands (one of SEEK(10), PRE\-FETCH(10) or
PRE\-FETCH(16)) that will be executed. The default value is 1. If an error
occurs it is noted and the program continues until \fINC\fR is exhausted.
If \fINC\fR is 0 then options are checked and the \fIDEVICE\fR is opened
but no commands are sent.
.TP
\fB\-g\fR, \fB\-\-grpnum\fR=\fIGN\fR
\fIGN\fR is the group number, a value between 0 and 63 (in hex: 0x3f). The
default value is 0. This option is ignored if the selected command is
SEEK(10).
.TP
\fB\-h\fR, \fB\-\-help\fR
output the usage message then exit.
.TP
\fB\-i\fR, \fB\-\-immed\fR
this option only applies to PRE\-FETCH(10) and PRE\-FETCH(16), setting
the IMMED bit. Without this option, the \fIDEVICE\fR returns after it has
completed transferring all, or part of, the requested blocks into the
cache. If this option is given the \fIDEVICE\fR returns after it has done
sanity checks on the cdb (e.g. making sure the \fILBA\fR is greater than
the number of available blocks) and before it does the transfer into the
cache.
.br
Note that even when this option is given, the return status from the
PRE\-FETCH commands is still either CONDITION MET status (if the cache seems
to have enough free space for the transfer) or a GOOD status (if the cache
does not seem to have enough free space).
.TP
\fB\-l\fR, \fB\-\-lba\fR=\fILBA\fR
\fILBA\fR is the starting logical block address that is placed in the
command descriptor block (cdb) of the selected command. Note that the
\fILBA\fR field in SEEK(10) and PRE\-FETCH(10) is a 32 bit quantity,
while with PRE\-FETCH(16) it is a 64 bit quantity. The default value is
0 .
.TP
\fB\-n\fR, \fB\-\-num\-blocks\fR=\fINUM\fR
\fINUM\fR is the number of blocks, starting at and including \fILBA\fR,
to place in the \fIDEVICE\fR's cache. The SEEK(10) command does not use
the \fINUM\fR value. For PRE\-FETCH(10) \fINUM\fR is a 16 bit quantity,
while for PRE\-FETCH(16) it is a 32 bit quantity. The default value is
1 . If \fINUM\fR is 0 then the \fIDEVICE\fR will attempt to transfer all
blocks from the given \fILBA\fR to the end of the medium.
.TP
\fB\-p\fR, \fB\-\-pre\-fetch\fR
this option selects either PRE\-FETCH(10) or PRE\-FETCH(16) commands. With
the \fI\-\-10\fR also given, the PRE\-FETCH(10) command is selected; without
that option PRE\-FETCH(16) is selected. The default (in the absence of this
and other 'selecting' options) the SEEK(10) command is selected.
.TP
\fB\-r\fR, \fB\-\-readonly\fR
this option sets a 'read\-only' flag when the underlying operating system
opens the given \fIDEVICE\fR. This may not work since operating systems can
not easily determine whether a pass\-through is a logical read or write
operation so they take a risk averse stance and require read\-write type
\fIDEVICE\fR opens irrespective of what is performed by the pass\-through.
.TP
\fB\-s\fR, \fB\-\-skip\fR=\fISB\fR
\fISB\fR is the number of logical block addresses to skip, between repeated
commands when \fINC\fR is greater than 1. The default value of \fISB\fR is
1 . \fISB\fR may be set to 0 so that all \fINC\fR PRE\-FETCH commands use
the same \fILBA\fR.
.TP
\fB\-t\fR, \fB\-\-time\fR
if given the elapsed time to execute \fINC\fR commands is recorded. This is
printed out before this utility exits. If \fINC\fR is greater than 1 then
the the "per command" time is also printed.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
increase the level of verbosity, (i.e. debug output).
.TP
\fB\-V\fR, \fB\-\-version\fR
print the version string and then exit.
.TP
\fB\-w\fR, \fB\-\-wrap\-offset\fR=\fIWO\fR
\fIWO\fR is the number of blocks, relative to \fILBA\fR, that when exceeded,
set the next command's logical block address back to \fILBA\fR. Whether
this "reset\-to\-LBA" action occurs depends on the values \fINC\fR and
\fISB\fR.
.SH NOTES
Prior to Linux kernel 4.17 the CONDITION MET status was logged as an error.
Recent versions of FreeBSD handle the CONDITION MET status properly.
.PP
If either the \fI\-\-count=NC\fR or \fI\-\-verbose\fR option is given then
a summary line like the following is output:
.PP
    Command count=5, number of condition_mets=3, number of goods=2
.PP
before the utility exits.
.SH EXIT STATUS
The exit status of sg_seek is 0 (GOOD) or 25 (CONDITION_MET) when this
utility is successful. If multiple commands are executed (e.g. when \fINC\fR
is greater than 1) then the result of the last executed SEEK or PRE\-FETCH
command sets the exit status. Otherwise see the sg3_utils(8) man page.
.SH AUTHORS
Written by Douglas Gilbert.
.SH "REPORTING BUGS"
Report bugs to <dgilbert at interlog dot com>.
.SH COPYRIGHT
Copyright \(co 2018 Douglas Gilbert
.br
This software is distributed under a BSD\-2\-Clause license. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B sg_vpd(sg3_utils); sdparm(sdparm)


CHANGELOG
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. All utilities in the main
directory have their own "man" pages. There is also a sg3_utils man page.

Changelog for sg3_utils-1.25 [20070710]
  - sg_dd: add oflag=sparse to step over bs*bpt number of zeros;
    - with oflag=sparse, write last bs*bpt segment at end or after
      error so file length of OFILE is appropriate
    - allow iflag=null and oflag=null which do nothing (placeholders)
    - when coe>1 then SCSI READ LONG logic remember extended block
      length of first encountered error
  - sg_ses: sync with ses2r17
  - place source in subversion repository

Changelog for sg3_utils-1.24 [20070507]
  - sg_raw: new utility to send arbirary SCSI commands
  - sg_luns: increase number of luns that can be fetched
    - fix length of raw and hex output
    - add '--quiet' option to output only ASCII hex
      representation of each lun
  - sg_rtpg: update for changes in spc4r09
  - sg_persist: update documentation, spc-4 references
    - fix exit status values
  - sg_inq: update version descriptors per spc4r09
    - fix '--id' and '--extended'
    - extend block limits VPD page (sbc3r09)
  - sg_vpd: extend block limits VPD page (sbc3r09)
    - append relative target port identifier to SAS target
      port address with '-iq' option
  - sg_logs: add decoding for stats+performance log pages
    - fix showing of page names for pdt > 0
    - implement '-HH' for single and all pages, fix '-r'
    - when '--maxlen=' given, only do single fetch
    - add Tape Alert (ssc), Media and Element statistics (smc) pages 
    - add '--brief' option
  - sg_ses: sync with ses2r16
    - fix bay number for SAS
  - sg_format: add '--dcrt' and '--security' options
  - sgm_dd: add 'smmap' oflag for shared_mmap_io testing
    - add 'dio' oflag
  - sg_dd, sgp_dd: add 'dio' iflag and oflag
  - sg_modes: change SAS mode page names per sas2r09
    - check validity of block descriptors length
  - sg_pt: change opaque context object from 'void *'
    to 'struct sg_pt_base *'
  - sg_opcodes: anticipate extra tmfs from 07-159r0
  - sg_sat_set_features: add more usage information
    - add man page
  - sg_sat_phy_event: add to examples directory
  - sg_lib: sync asc/ascq codes with spc4r10
  - Solaris port: using uscsi interface

Changelog for sg3_utils-1.23 [20070131]
  - sg_read_buffer: new utility
  - sg_write_buffer: new utility
  - sg_opcodes, sg_senddiag, sg_logs, sg_modes, sg_start, sg_inq,
    sg_turs, sg_readcap, sg_rbuf: add getopt_long() based cli;
    old and new cli selectable, new getopt_long cli is default
  - scripts: new subdirectory containing some bash scripts
    - add scripts/README file
  - sg_reassign: add '--hex' option for grown and primary lists
  - sg_rtpg: add '--raw' option
  - sg_lib.h, sg_cmds_basic.h + sg_cmds_extra.h: add C++ 
    'extern "C" ' wrappers
    - cleanup C code so it will compile as C++
  - sg_lib: sync with spc4r08
    - include <inttypes.h>, use PRId64 instead of %lld form
    - fix sg_get_sense_str() when empty sense buffer
  - win32 port: add Makefile.mingw + related support for MinGW
  - sg_cmds_extra: add sg_ll_read_buffer() and sg_ll_write_buffer()
  - sg_dd, sgp_dd, sgm_dd, sg_read: use lseek64() instead of llseek.c
  - sgm_dd: accept coe=<n> for interworking with sg_dd
  - sg_rdac: fix on non-linux ports
  - sg_ses: fix spurious warning in additional element status page
    - '-rr' option outputs a diagnostic page in binary to stdout
  - sg_opcodes: add command timeout descriptor support (spc4r08)
    - change linux specific pass through to generic pass through
  - sg_logs: add 'name=value' decoding for SAS specific lpage
  - examples+utils subdirectories: remove symlinks
  - synchronize man pages with usage messages
  - sg3_utils.spec: rework

Changelog for sg3_utils-1.22 [20061016]
  - sgp_dd: accept verbose=<n> as well as deb=<n> to ease
    interworking with sg_dd and sgm_dd
  - sg_sat_set_features: added to examples directory
  - sg_lib: sync asc/ascq text with spc4r06
    - move SG_LIB_CAT_NO_SENSE and SG_LIB_CAT_RECOVERED to
      20 and 21 respectively; add SG_LIB_CAT_ABORTED_COMMAND
      at 11 (its sense key value) 
  - sg_vpd: tweak '--page=sp --quiet' output
    - change '-HHH' so same as '-rr' (prepares ATA Information
      (ai) response for hdparm)
  - sg_requests: add '-s' option to set exit status from
    parameter data
  - sg_modes: exit quickly from '-e' if device not ready
  - sg_logs: sync sas log pages with sas2r05a
    - expand background scan results log page
    - add '-m=<max_len>' to limit response length
    - drop '-scum' and '-sthr' options and add '-select'
  - sg_write_long: add '--16' option to send 16 byte cdb
    - add '--wr_uncor' and '--pblock' options
  - sg_senddiag: cleanup and add sdiag_sas_p1_stop.txt
    to examples directory
  - sg_format: add '--cmplst=<n>' option (default: 1)
    - add '--pfu=<n>' option
    - expand man page to discuss P/D/C/GLISTs
  - sg_reassign: add '--primary' option to fetch primary
    defect list (PLIST) length
  - sg_readcap: add '-H' option to output response in hex
    and '-r' to output response in binary to stdout
    - add logical blocks per physical block (sbc3r07)
  - sginfo: add PLIST and GLIST designation to defect lists
  - sg_cmds: split this support file into sg_cmds_basic.[hc]
    and sg_cmds_extra.[hc]
    - add sg_ll_ata_pt() (SATL ATA pass) to sg_cmds_extra.[hc]
  - sg_rdac: fix includes for FreeBSD
  - sg_dd: add 'coe_limit=' option to exit after <n>
    consecutive 'coe' type read errors
  - sgm_dd: print out throughput information when signal arrives
    if time=1 (like sg_dd does already)
  - sg_inq: change '-HHH' so same as '-rr'. Now sg_inq, sg_vpd
    and sdparm output for hdparm with '-HHH'
    -add '-l=<resp_len>' option
  - sg_read_long: add '--pblock' option for physical blocks
  - sg_luns: add '--hex' and '--raw' options

src/sg_seek.c
/*
 * Copyright (c) 2018-2020 Douglas Gilbert.
 * All rights reserved.
 * Use of this source code is governed by a BSD-style
 * license that can be found in the BSD_LICENSE file.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>
#include <getopt.h>
#define __STDC_FORMAT_MACROS 1
#include <inttypes.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#if defined(HAVE_CLOCK_GETTIME) && defined(CLOCK_MONOTONIC)
#include <time.h>
#elif defined(HAVE_GETTIMEOFDAY)
#include <time.h>
#include <sys/time.h>
#endif

#include "sg_lib.h"
#include "sg_lib_data.h"
#include "sg_cmds_basic.h"
#include "sg_cmds_extra.h"
#include "sg_unaligned.h"
#include "sg_pr2serr.h"

/*
 * This program issues one or more SCSI SEEK(10), PRE-FETCH(10) or
 * PRE-FETCH(16) commands. Both PRE-FETCH commands are current and appear
 * in the most recent SBC-4 draft (sbc4r15.pdf at time of writing) while
 * SEEK(10) has been obsolete since SBC-2 (2004). Currently more hard disks
 * and SSDs support SEEK(10) than PRE-FETCH. It is even unclear what
 * SEEK(10) means (defined in SBC-1 as moving the hard disk heads to the
 * track containing the given LBA) for a SSD. But if the manufacturers'
 * support it, then it must have a use, presumably to speed the next access
 * to that LBA ...
 */

static const char * version_str = "1.08 20200115";

#define BACKGROUND_CONTROL_SA 0x15

#define CMD_ABORT_TIMEOUT  60      /* 60 seconds */


static struct option long_options[] = {
        {"10", no_argument, 0, 'T'},
        {"count", required_argument, 0, 'c'},
        {"grpnum", required_argument, 0, 'g'},
        {"help", no_argument, 0, 'h'},
        {"immed", no_argument, 0, 'i'},
        {"lba", required_argument, 0, 'l'},
        {"num-blocks", required_argument, 0, 'n'},
        {"num_blocks", required_argument, 0, 'n'},
        {"pre-fetch", no_argument, 0, 'p'},
        {"pre_fetch", no_argument, 0, 'p'},
        {"readonly", no_argument, 0, 'r'},
        {"skip", required_argument, 0, 's'},
        {"time", required_argument, 0, 't'},
        {"verbose", no_argument, 0, 'v'},
        {"version", no_argument, 0, 'V'},
        {"wrap-offset", required_argument, 0, 'w'},
        {"wrap_offset", required_argument, 0, 'w'},
        {0, 0, 0, 0},
};


static void
usage()
{
    pr2serr("Usage: "
            "sg_seek  [--10] [--count=NC] [--grpnum=GN] [--help] [--immed]\n"
            "                [--lba=LBA] [--num-blocks=NUM] [--pre-fetch] "
            "[--readonly]\n"
            "                [--skip=SB] [--time] [--verbose] [--version]\n"
            "                [--wrap-offset=WO] DEVICE\n");
    pr2serr("  where:\n"
            "    --10|-T             do PRE-FETCH(10) command (def: "
            "SEEK(10), or\n"
            "                        PRE-FETCH(16) if --pre-fetch also "
            "given)\n"
            "    --count=NC|-c NC    NC is number of commands to execute "
            "(def: 1)\n"
            "    --grpnum=GN|-g GN    GN is group number to place in "
            "PRE-FETCH\n"
            "                         cdb; 0 to 63 (def: 0)\n"
            "    --help|-h           print out usage message\n"
            "    --immed|-i          set IMMED bit in PRE-FETCH command\n"
            "    --lba=LBA|-l LBA    starting Logical Block Address (LBA) "
            "(def: 0)\n"
            "    --num-blocks=NUM|-n NUM    number of blocks to cache (for "
            "PRE-FETCH)\n"
            "                               (def: 1). Ignored by "
            "SEEK(10)\n");
    pr2serr("    --pre-fetch|-p     do PRE-FETCH command, 16 byte variant if "
            "--10 not\n"
            "                       given (def: do SEEK(10))\n"
            "    --readonly|-r      open DEVICE read-only (if supported)\n"
            "    --skip=SB|-s SB    when NC>1 skip SB blocks to next LBA "
            "(def: 1)\n"
            "    --time|-t          time the command(s) and if NC>1 show "
            "usecs/command\n"
            "                       (def: don't time)\n"
            "    --verbose|-v       increase verbosity\n"
            "    --version|-V       print version string and exit\n"
            "    --wrap-offset=WO|-w WO    if SB>0 and WO>0 then if "
            "LBAn>LBA+WO\n"
            "                       then reset LBAn back to LBA (def: 0)\n\n"
            "Performs SCSI SEEK(10), PRE-FETCH(10) or PRE-FETCH(16) "
            "command(s).If no\noptions are given does one SEEK(10) command "
            "with an LBA of 0 . If NC>1\nthen a tally is kept of successes, "
            "'condition-met's and errors that is\nprinted on completion. "
            "'condition-met' is from PRE-FETCH when NUM blocks\nfit in "
            "the DEVICE's cache.\n"
           );
}


int
main(int argc, char * argv[])
{
    bool cdb10 = false;
    bool count_given = false;
    bool do_time = false;
    bool immed = false;
    bool prefetch = false;
    bool readonly = false;
    bool start_tm_valid = false;
    bool verbose_given = false;
    bool version_given = false;
    int res, c;
    int sg_fd = -1;
    int first_err = 0;
    int last_err = 0;
    int ret = 0;
    int verbose = 0;
    uint32_t count = 1;
    int32_t l;
    uint32_t grpnum = 0;
    uint32_t k;
    uint32_t num_cond_met = 0;
    uint32_t num_err = 0;
    uint32_t num_good = 0;
    uint32_t numblocks = 1;
    uint32_t skip = 1;
    uint32_t wrap_offs = 0;
    int64_t ll;
    int64_t elapsed_usecs = 0;
    uint64_t lba = 0;
    uint64_t lba_n;
    const char * device_name = NULL;
    const char * cdb_name = NULL;
    char b[64];
#if defined(HAVE_CLOCK_GETTIME) && defined(CLOCK_MONOTONIC)
    struct timespec start_tm, end_tm;
#elif defined(HAVE_GETTIMEOFDAY)
    struct timeval start_tm, end_tm;
#endif

    while (1) {
        int option_index = 0;

        c = getopt_long(argc, argv, "c:g:hil:n:prs:tTvVw:", long_options,
                        &option_index);
        if (c == -1)
            break;

        switch (c) {
        case 'c':
            l = sg_get_num(optarg);
            if (l < 0) {
                pr2serr("--count= unable to decode argument, want 0 or "
                        "higher\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            count = (uint32_t)l;
            count_given = true;
            break;
        case 'g':
            l = sg_get_num(optarg);
            if ((l > 63) || (l < 0)) {
                pr2serr("--grpnum= expect argument in range 0 to 63\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            grpnum = (uint32_t)l;
            break;
        case 'h':
        case '?':
            usage();
            return 0;
        case 'i':
            immed = true;
            break;
        case 'l':
            ll = sg_get_llnum(optarg);
            if (-1 == ll) {
                pr2serr("--lba= unable to decode argument\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            lba = (uint64_t)ll;
            break;
        case 'n':
            l = sg_get_num(optarg);
            if (-1 == l) {
                pr2serr("--num= unable to decode argument\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            numblocks = (uint32_t)l;
            break;
        case 'p':
            prefetch = true;
            break;
        case 'r':
            readonly = true;
            break;
        case 's':
            l = sg_get_num(optarg);
            if (-1 == l) {
                pr2serr("--skip= unable to decode argument\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            skip = (uint32_t)l;
            break;
        case 't':
            do_time = true;
            break;
        case 'T':
            cdb10 = true;
            break;
        case 'v':
            verbose_given = true;
            ++verbose;
            break;
        case 'V':
            version_given = true;
            break;
        case 'w':
            l = sg_get_num(optarg);
            if (-1 == l) {
                pr2serr("--wrap-offset= unable to decode argument\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            wrap_offs = (uint32_t)l;
            break;
        default:
            pr2serr("unrecognised option code 0x%x ??\n", c);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }
    if (optind < argc) {
        if (NULL == device_name) {
            device_name = argv[optind];
            ++optind;
        }
        if (optind < argc) {
            for (; optind < argc; ++optind)
                pr2serr("Unexpected extra argument: %s\n",
                        argv[optind]);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }

#ifdef DEBUG
    pr2serr("In DEBUG mode, ");
    if (verbose_given && version_given) {
        pr2serr("but override: '-vV' given, zero verbose and continue\n");
        verbose_given = false;
        version_given = false;
        verbose = 0;
    } else if (! verbose_given) {
        pr2serr("set '-vv'\n");
        verbose = 2;
    } else
        pr2serr("keep verbose=%d\n", verbose);
#else
    if (verbose_given && version_given)
        pr2serr("Not in DEBUG mode, so '-vV' has no special action\n");
#endif
    if (version_given) {
        pr2serr("version: %s\n", version_str);
        return 0;
    }

    if (NULL == device_name) {
        pr2serr("Missing device name!\n\n");
        usage();
        return SG_LIB_SYNTAX_ERROR;
    }

    if (prefetch) {
        if (cdb10)
            cdb_name = "Pre-fetch(10)";
        else
            cdb_name = "Pre-fetch(16)";
    } else
        cdb_name = "Seek(10)";

    sg_fd = sg_cmds_open_device(device_name, readonly, verbose);
    if (sg_fd < 0) {
        if (verbose)
            pr2serr("open error: %s: %s %s\n", device_name, cdb_name,
                    safe_strerror(-sg_fd));
        ret = sg_convert_errno(-sg_fd);
        goto fini;
    }
#if defined(HAVE_CLOCK_GETTIME) && defined(CLOCK_MONOTONIC)
    if (do_time) {
        start_tm.tv_sec = 0;
        start_tm.tv_nsec = 0;
        if (0 == clock_gettime(CLOCK_MONOTONIC, &start_tm))
            start_tm_valid = true;
        else
            perror("clock_gettime(CLOCK_MONOTONIC)\n");
    }
#elif defined(HAVE_GETTIMEOFDAY)
    if (do_time) {
        start_tm.tv_sec = 0;
        start_tm.tv_usec = 0;
        gettimeofday(&start_tm, NULL);
        start_tm_valid = true;
    }
#else
    start_tm_valid = false;
#endif

    for (k = 0, lba_n = lba; k < count; ++k, lba_n += skip) {
        if (wrap_offs && (lba_n > lba) && ((lba_n - lba) > wrap_offs))
            lba_n = lba;
        res = sg_ll_pre_fetch_x(sg_fd, ! prefetch, ! cdb10, immed, lba_n,
                                numblocks, grpnum, 0, (verbose > 0), verbose);
        ret = res;      /* last command executed sets exit status */
        if (SG_LIB_CAT_CONDITION_MET == res)
            ++num_cond_met;
        else if (res) {
            ++num_err;
            if (0 == first_err)
                first_err = res;
            last_err = res;
        } else
            ++num_good;
    }

#if defined(HAVE_CLOCK_GETTIME) && defined(CLOCK_MONOTONIC)
    if ((count > 0) && start_tm_valid &&
        (start_tm.tv_sec || start_tm.tv_nsec)) {
        int err;

        res = clock_gettime(CLOCK_MONOTONIC, &end_tm);
        if (res < 0) {
            err = errno;
            perror("clock_gettime");
            if (EINVAL == err)
                pr2serr("clock_gettime(CLOCK_MONOTONIC) not supported\n");
        }
        elapsed_usecs = (end_tm.tv_sec - start_tm.tv_sec) * 1000000;
        /* Note that (end_tm.tv_nsec - start_tm.tv_nsec) may be negative */
        elapsed_usecs += (end_tm.tv_nsec - start_tm.tv_nsec) / 1000;
    }
#elif defined(HAVE_GETTIMEOFDAY)
    if ((count > 0) && start_tm_valid &&
        (start_tm.tv_sec || start_tm.tv_usec)) {
        gettimeofday(&end_tm, NULL);
        elapsed_usecs = (end_tm.tv_sec - start_tm.tv_sec) * 1000000;
        elapsed_usecs += (end_tm.tv_usec - start_tm.tv_usec);
    }
#endif

    if (elapsed_usecs > 0) {
        if (elapsed_usecs > 1000000)
            snprintf(b, sizeof(b), " (over %d seconds)",
                    (int)elapsed_usecs / 1000000);
        else
            b[0] = '\0';
        printf("Elapsed time: %" PRId64 " microseconds%s, per command time: "
               "%" PRId64 "\n", elapsed_usecs, b, elapsed_usecs / count);
    }

    if (count_given && verbose_given)
        printf("Command count=%u, number of condition_mets=%u, number of "
               "goods=%u\n", count, num_cond_met, num_good);
    if (first_err) {
        bool printed;

        printf(" number of errors=%d\n", num_err);
        printf("    first error");
        printed = sg_if_can2stdout(": ", first_err);
        if (! printed)
            printf(" code: %d\n", first_err);
        if (num_err > 1) {
            printf("    last error");
            printed = sg_if_can2stdout(": ", last_err);
            if (! printed)
                printf(" code: %d\n", last_err);
        }
    }
fini:
    if (sg_fd >= 0) {
        res = sg_cmds_close_device(sg_fd);
        if (res < 0) {
            pr2serr("close error: %s\n", safe_strerror(-res));
            if (0 == ret)
                ret = sg_convert_errno(-res);
        }
    }
    if (0 == verbose) {
        const char * e_str = (SG_LIB_CAT_CONDITION_MET == ret) ?
                             "sg_seek: " : "sg_seek: failed";
        
        if (! sg_if_can2stderr(e_str, ret))
            pr2serr("Some error occurred, try again with '-v' "
                    "or '-vv' for more information\n");
    }
    return (ret >= 0) ? ret : SG_LIB_CAT_OTHER;
}
